magenta-lib/src/main/scala/magenta/graph/Graph.scala

```diff
@@ -66,6 +96,50 @@ case class Graph[T](edges: Set[Edge[T]]) {
     Graph(newEdges)
   }
 
+  /**
+    * Flat map keeps the shape of this graph whilst mapping each node to a graph. This means one can replace the Start,
+    * End or Mid nodes with multiple nodes. The incoming and outgoing edges of each node of this graph will be replaced
+    * by edges that map to the start and end nodes of the graph returned from f.
+    *
+    * An identity mapping for flatMap would be start and end nodes to empty map and mid nodes to a single value graph
+    * containing the value of the node.
+    *
+    * It is not possible to map a mid node to an empty node.
+    * @param f the function to apply to each element
+    * @return a new graph with the graphs substituted
+    */
+  def flatMap[R](f: Node[T] => Graph[R]): Graph[R] = {
+    val nodeToGraphMap: Map[Node[T], Graph[R]] = nodes.map(n => n -> f(n)).toMap
+    // find all of the edges that we won't adjoin to another graph
+    val allInternalEdges = nodeToGraphMap.flatMap{
+      case (StartNode, graph) => graph.edges -- graph.incoming(EndNode)
+      case (MidNode(_), graph) =>
+        assert(!graph.isEmpty, "It is not possible to flatMap a MidNode to an empty graph")
+        graph.edges -- graph.incoming(EndNode) -- graph.outgoing(StartNode)
+      case (EndNode, graph) => graph.edges -- graph.outgoing(StartNode)
+    }.toSet
+
+    // sort the edges by priority - this means we can re-calculate the priorities correctly
+    val joiningEdges = edges.toList.sortBy(_.priority)
+      .foldLeft(Set.empty[Edge[R]]) { case(acc, oldEdge) =>
+
+      val fromGraph = nodeToGraphMap(oldEdge.from)
+      val toGraph = nodeToGraphMap(oldEdge.to)
+      // get the joining edges
+      val newEdges = Graph.joiningEdges(fromGraph, toGraph)
+        // deduplicate any edges that we've already generated
+        .filterNot(e1 => acc.exists(e1 ~= _))
+        // fix up the priorities
+        .map{ newEdge =>
+          val maxPriority = acc.filter(_.from == newEdge.from).map(_.priority).reduceOption(Math.max).getOrElse(0)
+          newEdge.copy(priority = newEdge.priority + maxPriority)
```

Comment:
Reprioritise assumes we have all of the edges to reprioritise. In this case we only have the new edges that are the result of processing one old edge. Other edges that leave the same node can already have been accumulated - in which case we use the max priority to offset the existing priorities.

It might be possible to reuse the reprioritise call but I think the code would be more complex (we have to sort the sets into lists first).